packages/react/__tests__/rehydration.js

```diff
@@ -166,3 +197,131 @@ test('initializing another Emotion instance should not move already moved styles
     </head>
   `)
 })
+
+test('xxx', () => {
+  const { app, styles } = disableBrowserEnvTemporarily(() => {
+    resetAllModules()
+
+    let cache = createCache({ key: 'mui' })
+    let { extractCritical2, constructStyleTags } = createEmotionServer(cache)
+
+    const rendered = ReactDOMServer.renderToString(
+      <CacheProvider value={cache}>
+        <Global styles={{ body: { color: 'white' } }} />
+        <Global styles={{ html: { background: 'red' } }} />
+        <main css={{ color: 'green' }}>
+          <div css={{ color: 'hotpink' }} />
+        </main>
+      </CacheProvider>
+    )
+    const extracted = extractCritical2(rendered)
+    return {
+      app: extracted.html,
+      styles: constructStyleTags(extracted)
+    }
+  })
+
+  safeQuerySelector('head').innerHTML = styles
+  safeQuerySelector('body').innerHTML = `<div id="root">${app}</div>`
+  expect(safeQuerySelector('html')).toMatchInlineSnapshot(`
+    <html>
+      <head>
+        <style
+          data-emotion="mui-global l6h"
+        >
+          body{color:white;}
+        </style>
+        <style
+          data-emotion="mui-global 10q49a4"
+        >
+          html{background:red;}
+        </style>
+        <style
+          data-emotion="mui bjcoli 1lrxbo5"
+        >
+          .mui-bjcoli{color:green;}.mui-1lrxbo5{color:hotpink;}
+        </style>
+      </head>
+      <body>
+        <div
+          id="root"
+        >
+          <main
+            class="mui-bjcoli"
+          >
+            <div
+              class="mui-1lrxbo5"
+            />
+          </main>
+        </div>
+      </body>
+    </html>
+  `)
+
+  resetAllModules()
+  const cache = createCache({ key: 'mui' })
+
+  ReactDOM.render(
+    <CacheProvider value={cache}>
+      <Global styles={{ body: { color: 'white' } }} />
+      <Global styles={{ html: { background: 'red' } }} />
+      <main css={{ color: 'green' }}>
+        <div css={{ color: 'hotpink' }} />
+      </main>
+    </CacheProvider>,
+    safeQuerySelector('#root')
+  )
+
+  expect(safeQuerySelector('head')).toMatchInlineSnapshot(`
+    <head>
+      <style
+        data-emotion="mui-global"
+        data-s=""
+      >
+        
+        body{color:white;}
+      </style>
+      <style
+        data-emotion="mui-global"
+        data-s=""
+      >
+        
+        html{background:red;}
+      </style>
```

Comment:
SSRed global style elements were replaced here by client-rendered ones, this all happens very quickly so this might not be noticeable (although I'm slightly worried about unmounting `@font-face` and stuff like that, even for a brief moment) but I would still try to check what could be done about it

Why does this happen? SSR style is immediately flushed here:
https://github.com/emotion-js/emotion/blob/4d7efcb33c5e70f6f1732a96023a25f5270e72dd/packages/react/src/global.js#L126-L131
since we already have some rehydrated `sheet.tags` coming from here: 
https://github.com/emotion-js/emotion/blob/4d7efcb33c5e70f6f1732a96023a25f5270e72dd/packages/react/src/global.js#L108-L110

I believe that the original intention was to flush this on updates but right now we are also flushing it when rehydrating 😕

Reply:
Ah, this may be the reason for the flickering here - https://deploy-preview-25690--material-ui.netlify.app/ The text is black for a split second, then it changes to white, which is coming from the global styles. Although on this example, I am just omitting the global styles coming from `extractCritical2`, so it may be a different issue. Will have to double-check.

Reply:
Hm, I can only see the `color:white` declaration in the SSRed HTML 🤔 Has this site been updated since then?

Reply:
I cannot spot the flickering too this one time.. I’ve updated the example, we should see the borders after 5s too as that Global component stays mounted. Only the color should change to black (default)

Reply:
@mitchellhamilton I would like to fix this issue with flushing and reinjecting Global styles when rehydrating. The easiest fix for that would be to just merge those 2 layout effects. Any objections to that? The first effect just avoids recreating StyleSheet (from what I can tell) based on the `serialized.name` but since `sheet.flush` is called anyway when it changes and highly dynamic global styles are rather rare I don't think we must maintain that split between two effects 🤔

Reply:
Sounds good

Reply:
I've implemented this change - could you take a look @mitchellhamilton ? I doubt that this can break anything but better to recheck this

Reply:
Hmmm, reading the code, it seems as though the change doesn't preserve the behavior of inserting the new style tag in the same position as the previous one. Could you add a test based on https://codesandbox.io/s/still-leaf-invs2?file=/src/App.js?